### ISSUE TITLE
Consider passing custom attributes to components

### ISSUE THREAD

**Issue by adamkudrna** [react-ui-org/react-ui#14]
Currently, most components cannot have custom attributes like `className`, `aria-label` or similar. It might be useful if they could — for example, adding CSS utility classes for offsets would be easier. Optional attributes could be of any type, they would not be defined in component's props.

## Passing CSS class
Current state:
```jsx
<div className="mb-3">
  <Button label="Button with classname" />
</div>
```

~~Proposal:~~
⚠️ This is not a desired solution, it could encourage people to hack the component's inner styles.
```jsx
<Button className="mb-3" label="Button with classname" />
```

👉 Proposal: special utility props, see #18.

## Passing eg. ARIA attribute for accessibility
Proposal (currently supported only in `ButtonGroup` and `CardList`):
```jsx
<ButtonGoup aria-label="group">
  …
</ButtonGroup>
```

**Comment by bedrich-schindler** [comment]
After personal discussion we agreed on passing native attributes without proptype checking. So we need to remove unnecessary prop types (like autocapitalize, autocomplete) and decide way how to pass native attributes to component.

There are two possible ways to pass native attributes to components:

1. Pass all attributes directly as props using spread operator (Material UI use follow this way)
2. Pass all native attributes via special prop type (e.x. `nativeAttributes`):

```
<TextField
  changeHandler={() => {}}
  label="Username"
  nativeAttributes={{
    autoComplete: true,
    autoCapitalize: false,
  }}
/>
```

@mbohal @adamkudrna What do you think?

**Comment by mbohal** [comment]
I would prefer option 2. I think I would prefer calling it `htmlAttibutes` though. Im affraid `nativeAttributes` could wrongly make it seem related to React Native or such.

**Comment by adamkudrna** [comment]
I'm for 2 named as `htmlAttributes`.

How about custom class names (the first point of the issue)? Do we want to allow that?

**Comment by bedrich-schindler** [comment]
It depends on component, for example top-level element of `Button` component is `button`. Top-level element of `TextField` component is `div`, not `input`. 

Is is hard to decide but I my opinion, I aim to be able to change attributes of native element, it means that we want to pass attributes like this:

- `Button` -> `button`
- `ButtonGroup` -> `button` (passed to each button)
- `Checkbox` -> `input`
- `MultipleSelectField` -> `select`
- `Radio` -> `input` 
- `SelectField` -> `select`
- `TextField` -> `input`
- `TextArea` -> `textarea`
- `Toggle` -> `input`

We might have been talking about two separate things, you aim to be able to set classnames of inner elements, but our goal (with @mbohal) of this issue is to able to change attributes like `max-lenght`, `min-lenght`, `autocomplete`, `autocapture` and attributes like this.

These two points of view are little bit different. This is the reason I want to call it `nativeElementAttributes` because we need to pass attributes to native elements that are listed above.

Another part is to pass classnames to inner elements and it is not so easy because each element consists of different number of elements. Sometimes there is label, sometimes there is span. There can be 4 elements, but also 10 elements. Me personally, I would expect to be able to set only classname of top-level component because I don`t know inner representation of component and only think that I often need, is to set margin, or overwrite width or thinks like things, I have never need to change appearance of inner part of component - if it is needed, you need to write your own component - If you remember CES project, we ran into the hell with customizing MUI elements and we wrote our components in the end.

As I think about it, my `nativeElementAttributes` or whather it is called, are only for components listed above, whereby `className` can be passed to every component. If it changes top-level component, I would call it just `className`.

**Comment by adamkudrna** [comment]
Yes, in general I do agree. To structure the topics:

1. Yes, let's do it for listed components as you suggest.
2. `nativeElementAttributes` is a good name when I know what you mean but I'm afraid that @mbohal's objection is [still valid](https://react-native-elements.github.io/react-native-elements/)… Is there a better name…? Something common to inputs and buttons? Maybe `controlHtmlAttributes`?
3. Aside from 1., I believe there HTML attributes (not talking about class names this time) we may want to pass to the topmost HTML element of the component. We already do it with `ButtonGroup` to improve accessibility (`aria-label`):

```jsx
<ButtonGroup
  aria-label={t('component:configFormType.table.text.viewMode')}
  disabled={invalidSeids.size > 0}
>
```
4. I don't wish to pass custom class names into components and in #18 I explain why (corresponds to what you are saying).

**Comment by bedrich-schindler** [comment]
OK, `nativeElementAttributes` seems to be crystal clear to me but I agree that I can colide with library you linked above. 

But I am still woried about these props for passing custom attributes into component. As I wrote above, sometimes top-level components is equal to HTML element and sometimes don't. It can result in perplexity if there is somethink like `controlHtmlAttributes` and `topLevelHtmlAttributes` is some cases and only `topLevelHtmlAttributes` in other case.

`topLevelHtmlAttributes` will be present there everytime I think, Am I correct? (I does not matter what name is used, it is only example - bad example). And `controlHtmlAttributes` only for components above, Am I correct? What about if top-level component is html element in same time, only `topLevelHtmlAttributes`  is present then? Will not be user confused that it is missing?

**Comment by mbohal** [comment]
In reaction to https://github.com/react-ui-org/react-ui/issues/14#issuecomment-595727645

1. I agree
2. How about `htmlElementAttributes`?
3. I think that would make the API too way complex. The idea is essentiaally being able to add any attributes to anything inside a component. I think we should not try to support this as it would make the dev experience poor. If we need aria attributes, lets make set them explicitly with props on the component itself. I think I agree with @bedrich-schindler on this one.

**Comment by mbohal** [comment]
@adamkudrna  
3. You are right, my last suggestion would bring inconsistency. I think we should go with `topLevelHtmlElementAttributes`. Although I understand what @bedrich-schindler is saying in https://github.com/react-ui-org/react-ui/issues/14#issuecomment-595739199 I think it is the best solution. Everything else I can think of is much worse.

**Comment by adamkudrna** [comment]
Based on personal discussion: we decided to implement only `htmlElementAttributes` at the moment. We may get back to `topLevelHtmlElementAttributes` later.

**Comment by bedrich-schindler** [comment]
Last time we discussed it, we decided that is it not a good idea to do it this way. We will change add missing props to each component that need them.

**Comment by bedrich-schindler** [comment]
Via mobile decided that we will use spread operator to pass other props to component.
